### PR TITLE
AdamaxOpt: Adds Adamax Optimizer with eager mode

### PR DESCRIPTION
**Pull request by ManrajGrover**
This PR adds Adamax Optimizer with eager mode

---
This change is [Reviewable](https://reviewable.io/reviews/pair-code/deeplearnjs/691)

### CODE REVIEW

**Comment by nsthorat** [comment]
Review status: 0 of 7 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[src/math/optimizers/adamax_optimizer.ts, line 45 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/691#-L55M0SRxjzymTajfktl:-L55M0SRxjzymTajfktm:b-zdhb8q) ([raw file](https://github.com/pair-code/deeplearnjs/blob/99d61934e1ae7fed50e008022549dbc69f253416/src/math/optimizers/adamax_optimizer.ts#L45)):*
> ```XML
>       protected learningRate: number, private beta1: number,
>       private beta2: number,
>       /** @deprecated */ specifiedVariableList?: Node[], epsilon = 1e-8) {
> ```

we're missing a decay here (to be TF compatible). It looks like this is just how much the learning rate decays at each time step.

https://www.tensorflow.org/api_docs/python/tf/keras/optimizers/Adamax

---

*[src/math/optimizers/adamax_optimizer.ts, line 50 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/691#-L55LkVNZ-Dqgfuu9GjY:-L55LkVOG8rl8zDlFtOe:b-do2u5u) ([raw file](https://github.com/pair-code/deeplearnjs/blob/99d61934e1ae7fed50e008022549dbc69f253416/src/math/optimizers/adamax_optimizer.ts#L50)):*
> ```XML
>     this.eps = keep(scalar(epsilon));
>     // b1, b2 keep initial value of beta* hyperparameters.
>     this.b1 = keep(scalar(this.beta1));
> ```

call this beta1, beta2, accBeta1, etc and make accB1 a variable

---

*[src/math/optimizers/adamax_optimizer.ts, line 99 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/691#-L55LfouyQFQY3j6oUec:-L55LfouyQFQY3j6oUed:b6qpct0) ([raw file](https://github.com/pair-code/deeplearnjs/blob/99d61934e1ae7fed50e008022549dbc69f253416/src/math/optimizers/adamax_optimizer.ts#L99)):*
> ```XML
>   }
> 
>   _disposeAndUpdateBeta() {
> ```

don't make this a function, just do it inside the tidy() above

Also, make the beta values variables.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pair-code/deeplearnjs/691)*

**Comment by nsthorat** [comment]
Review status: 0 of 7 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[src/math/optimizers/adamax_optimizer.ts, line 99 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/691#-L55LfouyQFQY3j6oUec:-L55NNh2uwcTw6-8Hwjq:b-ke6fk9) ([raw file](https://github.com/pair-code/deeplearnjs/blob/99d61934e1ae7fed50e008022549dbc69f253416/src/math/optimizers/adamax_optimizer.ts#L99)):*
<details><summary><i>Previously, nsthorat (Nikhil Thorat) wrote…</i></summary><blockquote>

don't make this a function, just do it inside the tidy() above

Also, make the beta values variables.
</blockquote></details>

just accB1 as a variable, to be clear.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pair-code/deeplearnjs/691)*

**Comment by ManrajGrover** [comment]
@nsthorat Actually, the learning rate decay here is something not mentioned in the [paper](https://arxiv.org/pdf/1412.6980v8.pdf). Should we still add it?

**Comment by nsthorat** [comment]
Yeah let's still add it to be TF compatible.

**Comment by nsthorat** [comment]
:lgtm_strong:

---

Review status: 0 of 6 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/pair-code/deeplearnjs/691#-:-L5JjaI_qZeBw0hE8M0h:b-jbsf8z)*

**Comment by nsthorat** [comment]
One more merge with master :)

---

Review status: 0 of 6 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/pair-code/deeplearnjs/691#-:-L5JjbXHrA5ZNSAsAIKz:b-awhuhl)*

**Comment by nsthorat** [comment]
Hey @ManrajGrover, can you run the unit tests? I changed something with dl.variable() which will make your tests fail. See what I had to do here:
https://github.com/PAIR-code/deeplearnjs/commit/0627b39e1910bd1c0fe7670f12a6fe65831636a7

**Comment by nsthorat** [comment]
:lgtm_strong:

---

Review status: 0 of 6 files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/pair-code/deeplearnjs/691#-:-L5Ox0TNLFscHC58lsuC:b-jbsf8z)*
